plugins/inputs/vsphere/vsan.go

```diff
@@ -0,0 +1,534 @@
+package vsphere
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25"
+	"github.com/vmware/govmomi/vim25/methods"
+	"github.com/vmware/govmomi/vim25/soap"
+	"github.com/vmware/govmomi/vim25/types"
+	vsanmethods "github.com/vmware/govmomi/vsan/methods"
+	vsantypes "github.com/vmware/govmomi/vsan/types"
+)
+
+const (
+	vsanNamespace    = "vsan"
+	vsanPath         = "/vsanHealth"
+	hwMarksKeyPrefix = "vsan-perf-"
+	perfPrefix       = "performance."
+)
+
+var (
+	vsanPerfMetricsName    string
+	vsanSummaryMetricsName string
+	perfManagerRef         = types.ManagedObjectReference{
+		Type:  "VsanPerformanceManager",
+		Value: "vsan-performance-manager",
+	}
+	hyphenReplacer = strings.NewReplacer("-", "")
+)
+
+// collectVsan is the entry point for vsan metrics collection
+func (e *Endpoint) collectVsan(ctx context.Context, resourceType string, acc telegraf.Accumulator) error {
+	lower, err := versionLowerThan(e.apiVersion, "5.5")
+	if err != nil {
+		return fmt.Errorf("Failed to get the vCenter version for vSAN: %v", err)
+	}
+	if lower {
+		return fmt.Errorf("A minimum API version of 5.5 is required for vSAN. Found: %s. Skipping VCenter: %s", e.apiVersion, e.URL.Host)
+	}
+	vsanPerfMetricsName = strings.Join([]string{"vsphere", "vsan", "performance"}, e.Parent.Separator)
+	vsanSummaryMetricsName = strings.Join([]string{"vsphere", "vsan", "summary"}, e.Parent.Separator)
+	res := e.resourceKinds[resourceType]
+	client, err := e.clientFactory.GetClient(ctx)
+	if err != nil {
+		return fmt.Errorf("fail to get client when collect vsan: %v", err)
+	}
+	// Create vSAN client
+	vimClient := client.Client.Client
+	vsanClient := vimClient.NewServiceClient(vsanPath, vsanNamespace)
+	// vSAN Metrics to collect
+	metrics := e.getVsanMetadata(ctx, vsanClient, res)
+	// Iterate over all clusters, run a goroutine for each cluster
+	te := NewThrottledExecutor(e.Parent.CollectConcurrency)
```

Comment:
Do we need to recreate those elements like clients etc. in every gather cycle? Why not simply create it at `Init()` and use it here?